src/Templating/Helper/BlockHelper.php

```diff
@@ -134,6 +138,18 @@ public function __construct(
         $this->blockContextManager = $blockContextManager;
         $this->cacheHandler = $cacheHandler;
         $this->stopwatch = $stopwatch;
+
+        // NEXT_MAJOR: remove this check
+        if (null !== $cacheManager) {
+            @trigger_error(
+                sprintf(
+                    'Passing an instance of "%s" as argument 6 for method "%s" is deprecated since sonata-project/block-bundle 4.x. The argument will be removed in 5.0.',
```

Comment:
Since the argument will be removed, 
people shouldn't do
```
new BlockHelper($blockServiceManager, $cacheBlocks, $blockRenderer, $blockContextManager, $eventDispatcher, null, $cacheHandler, $stopwatch)
```
but
```
new BlockHelper($blockServiceManager, $cacheBlocks, $blockRenderer, $blockContextManager, $eventDispatcher, $cacheHandler, $stopwatch)
```
so we need to allow this syntax if we want to provide a BC upgrade.

Also, `cacheBlocks` and `cacheHandler` need to be deprecated. So the new signature will be
```
new BlockHelper($blockServiceManager, $blockRenderer, $blockContextManager, $eventDispatcher, $stopwatch)
```

So we need something like
```
public function __construct(
        BlockServiceManagerInterface $blockServiceManager,
        $blockRendererOrCacheBlocks,
        object $blockContextManagerOrBlockRenderer,
        object $eventDispatcherOrBlockContextManager,
        ?object $stopwatchOrEventDispatcher = null,
        ?CacheManagerInterface $cacheManager = null,
        ?HttpCacheHandlerInterface $cacheHandler = null,
        ?Stopwatch $stopwatch = null
    ) {
```
and the check can be done on `blockRendererOrCacheBlocks` if it's an array (deprecated) or a blockRenderer.

Reply:
But then, for the service definition I wonder how to avoid the deprecation, should we have to service definition and using one or the other based on the config ?

Reply:
added a compiler pass to update to the new signature in case the array of cache blocks is empty